pipe.sh

```diff
@@ -1,19 +1,160 @@
-#!/usr/bin/env ash
+#!/usr/bin/env sh
 
 set -e
 
-composer --version
-composer create-project --repository-url="https://mirror.mage-os.org/" "magento/project-community-edition:>=2.4.5 <2.4.6" ./magento2 --no-install
-cd magento2
-composer config --no-interaction allow-plugins.dealerdirect/phpcodesniffer-composer-installer true
-composer config --no-interaction allow-plugins.laminas/laminas-dependency-plugin true
-composer config --no-interaction allow-plugins.magento/* true
-cat composer.json
-composer install
-cd dev/tests/integration
-sed -i "s/'db-host' => 'localhost'/'db-host' => '127.0.0.1'/" etc/install-config-mysql.php.dist
-sed -i "s/'db-user' => 'root'/'db-user' => 'user'/" etc/install-config-mysql.php.dist
-sed -i "s/'db-password' => '123123q'/'db-password' => 'password'/" etc/install-config-mysql.php.dist
-sed -i "s/'elasticsearch-host' => 'localhost'/'elasticsearch-host' => '127.0.0.1'/" etc/install-config-mysql.php.dist
-sed -i "s/'amqp-host' => 'localhost'/'amqp-host' => '127.0.0.1'/" etc/install-config-mysql.php.dist
-php ../../../vendor/bin/phpunit ../../../vendor/magento/magento2-base/dev/tests/integration/testsuite/Magento/Framework/MessageQueue/TopologyTest.php
+TYPE=${TYPE:="integration"}
+
+# Database defaults
+DATABASE_USERNAME=${DATABASE_USERNAME:="user"}
+DATABASE_ROOTPASSWORD=${DATABASE_ROOTPASSWORD:="rootpassword"}
+DATABASE_PASSWORD=${DATABASE_PASSWORD:="password"}
+
+# Service defaults
+ELASTICSEARCH_HOST=${ELASTICSEARCH_HOST:="host.docker.internal"}
+RABBITMQ_HOST=${RABBITMQ_HOST:="host.docker.internal"}
+DATABASE_HOST=${DATABASE_HOST:="host.docker.internal"}
+
+REPOSITORY_URL=${REPOSITORY_URL:="https://repo.magento.com/"}
+MAGENTO_VERSION=${MAGENTO_VERSION:="magento/project-community-edition:>=2.4.5 <2.4.6"}
+
+GROUP=${GROUP:=""}
+TESTS_PATH=${TESTS_PATH:=""}
+
+create_database_schema () {
+  mysql -h $DATABASE_HOST -uroot -p$DATABASE_ROOTPASSWORD << SQL
+  CREATE DATABASE IF NOT EXISTS $1;
+  CREATE USER IF NOT EXISTS '$DATABASE_USERNAME'@'%' IDENTIFIED BY '$DATABASE_PASSWORD';
+  GRANT ALL ON $1.* TO '$DATABASE_USERNAME'@'%';
+  FLUSH PRIVILEGES;
+SQL
+}
+
+run_integration_tests () {
+  if [ ! -f "composer.lock" ]; then
+    echo "composer.lock does not exist."
+    composer create-project --repository-url="$REPOSITORY_URL" "$MAGENTO_VERSION" ./magento2 --no-install
+    cd magento2
+  fi
+
+  composer config --no-interaction allow-plugins.dealerdirect/phpcodesniffer-composer-installer true
+  composer config --no-interaction allow-plugins.laminas/laminas-dependency-plugin true
+  composer config --no-interaction allow-plugins.magento/* true
+  cat composer.json
+  composer install
+  cd dev/tests/integration
```

Comment:
These 6 or so lines are duplicated a few times. Can we pull this into a `composer_config_setup()` function or something similar?

Reply:
Yep - the latest commit showed the need for this. Extra lines were added only to the graphql function and not the other 2

Reply:
thanks, i'll resolve this next

Reply:
this is resolved by #2